### PR TITLE
Consolidate output and input utxo behaviour

### PR DESCRIPTION
**Pull request by achiurizo**
:clipboard:  https://github.com/omisego/elixir-omg/pull/1212  larger scoped PR

## Overview

This is a first pass at integrating our [ex_plasma]() library to consolidate our encoding/decoding protocol with the plasma framework contracts. Currently, the codebase has begun to future code the ALD behavior. This reverts that and encourages the protocol related changes to be moved over to the library.

This is multi-stage integration. This first pass starts offloading the input utxo calculation and cleans up the structure to use more of `ex_plasma` data modeling. This also consolidates the output utxo modules a bit too. Following the discussion on #1212, this is intentionally a _smaller scope_ as we move towards the integration of said library.

Follow up tasks:

- [ ] refactor output utxo behavior/calculations and leverage `ExPlasma.Utxo`
- [ ] refactor transactions to use `ExPlasma.Transaction` , and it's other types like `Payment` and `Deposit`.
- [ ] port Fees transaction to `ExPlasma` and leverage that.
- [ ] Remove dispatching module code, like `WireTransactionTypes`. This behavior should exist in the library.

## Changes

* Consolidates:

  * `OMG.InputPointer`  - this module "*represents the data that's used to mention outputs that are intended to be inputs to a transaction*"
  * `OMG.InputPointer.UtxoPosition` this "*represents the utxo position interpreted as an implementation of an input pointer type*"
  * `OMG.InputPointer.Protocol` - this "*captures the varying behavior of how the outputs can be pointed to in child chain transactions*"

  into `OMG.Utxo.Position`, a pre-existing module that the modules above invoke in the call stack. This is a "*representation of a UTXO position in the child chain, providing encoding/decoding to/from formats digestible in 'Eth' and in the `OMG.DB`*".

  Consolidating it into `Utxo.Position` seems like the best first pass as all the above modules end up [calling functions](https://github.com/omisego/elixir-omg/compare/achiurizo/ex_utxo_2?expand=1#diff-64563a0f252706476c8cc34790e20564L30) in `Utxo.Position` [in their call stack](https://github.com/omisego/elixir-omg/compare/achiurizo/ex_utxo_2?expand=1#diff-417dc52abc6b8f820e190f8726f725d7L27). Once we've merged the module behaviors, we can start delegating the functionality to `ex_plasma` more. Currently there are some behaviors that need to be aligned and refactored(`OMG.Output` error tuple behavior is one such example)

* Adds [doctests/unit tests](https://github.com/omisego/elixir-omg/compare/achiurizo/ex_utxo_2?expand=1#diff-e24a6d0e13aa92e737647e776f74cadbR41) for `OMG.Utxo.Position`. We didn't have unit test coverage for `OMG.InputPointer` , `OMG.InputPointer.Position`, `OMG.InputPointer.Protocol`. 

* Consolidates:

  * `OMG.Output.Procotocol` - This "captures the varying behavior of outputs that build the plasma chain"
  * `OMG.Output.FungibleMoreVPToken` - This is a "representation of the payment transaction output of a fungible token `currency`"

  Into `OMG.Output`, which is "specifically dispatches generic calls to the various specific types". We currently have 1 type of output type so I think we should be wary of this future coding. Additionally, we should begin moving this encoding/decoding protocol behavior into `ex_plasma` to handle [the dispatching](https://github.com/omisego/ex_plasma/blob/master/lib/ex_plasma/transaction.ex#L134).

* Refactor functions that have similiar/exact behaviors and remove functions that were not used. Some examples:

  * [OMG.Utxo.Position.decode/1](https://github.com/omisego/elixir-omg/compare/achiurizo/ex_utxo_2?expand=1#diff-e24a6d0e13aa92e737647e776f74cadbR92).
  * `OMG.Utxo.Position.blknum/1`, `OMG.Utxo.Position.txindex/1`, `OMG.Utxo.Position.oindex/1`. These were at most used 1 time and a [pattern match](https://github.com/omisego/elixir-omg/compare/achiurizo/ex_utxo_2?expand=1#diff-5f4c08343920f8d7c53cf7e680c3f613R140) directly would have sufficed.
  * `OMG.Output.reconstruct/1`  - this merges with `OMG.Output.dispatching_reconstruct/1` as the behaviour is now merged.

## Testing

Run CI, alternatively, run the commands:

```
$ mix test
```

### CODE REVIEW

File: apps/omg/mix.exs [hunk 1]
```diff
@@ -30,6 +30,7 @@ defmodule OMG.MixProject do
 
   defp deps do
     [
+      {:ex_plasma, git: "https://github.com/omisego/ex_plasma.git"},
```
Comment (InoMurko) [inline apps/omg/mix.exs]: here it comeeesss 👏 👏 👏

File: apps/omg/lib/omg/utxo/position.ex [hunk 2]
```diff
@@ -19,8 +19,7 @@ defmodule OMG.Utxo.Position do
   """
 
   # these two offset constants are driven by the constants from the RootChain.sol contract
```
Comment (T-Dnzt) [inline apps/omg/lib/omg/utxo/position.ex]: Should probably remove this line. ☝️

File: apps/omg/lib/omg/utxo/position.ex [hunk 3]
```diff
@@ -39,50 +38,138 @@ defmodule OMG.Utxo.Position do
...
+  ## Examples
+
+      iex> utxo_pos = {:utxo_position, 4, 5, 1}
+      iex> OMG.Utxo.Position.encode(utxo_pos)
+      4_000_050_001
+  """
   @spec encode(t()) :: non_neg_integer()
   def encode(Utxo.position(blknum, txindex, oindex)) when is_position(blknum, txindex, oindex),
-    do: blknum * @block_offset + txindex * @transaction_offset + oindex
+    do: ExPlasma.Utxo.pos(%{blknum: blknum, txindex: txindex, oindex: oindex})
```
Comment (T-Dnzt) [inline apps/omg/lib/omg/utxo/position.ex]: 🎉

File: apps/omg/lib/omg/state/transaction/fee_token_claim.ex [hunk 4]
```diff
@@ -103,15 +103,15 @@ defimpl OMG.State.Transaction.Protocol, for: OMG.State.Transaction.FeeTokenClaim
   def get_data_for_rlp(%Transaction.FeeTokenClaim{tx_type: tx_type, outputs: outputs, nonce: nonce}) do
     [
       tx_type,
-      Enum.map(outputs, &OMG.Output.Protocol.get_data_for_rlp/1),
+      Enum.map(outputs, &OMG.Output.get_data_for_rlp/1),
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/transaction/fee_token_claim.ex]: two items of the list are variables, one of them isn't. lets make this into
```
rlp_outputs = Enum.map(outputs, &OMG.Output.get_data_for_rlp/1)
[
tx_type,
rlp_outputs,
nonce
]
```

File: apps/omg/lib/omg/state/transaction/payment.ex [hunk 5]
```diff
@@ -170,18 +168,18 @@ defimpl OMG.State.Transaction.Protocol, for: OMG.State.Transaction.Payment do
       when Transaction.is_metadata(metadata),
       do: [
         tx_type,
-        Enum.map(inputs, &OMG.InputPointer.Protocol.get_data_for_rlp/1),
-        Enum.map(outputs, &OMG.Output.Protocol.get_data_for_rlp/1),
+        Enum.map(inputs, &OMG.Utxo.Position.get_data_for_rlp/1),
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/transaction/payment.ex]: same as above, let's name these two mappings.

File: apps/omg/lib/omg/output.ex [hunk 6]
```diff
@@ -19,56 +19,75 @@ defmodule OMG.Output do
...
 
-  @doc """
-  True if a particular witness can unlock a particular output to be spent, given being put in a particular transaction
+  def from_db_value(%{type: output_type} = db_value), do: @output_types_modules[output_type].from_db_value(db_value)
 
-  Intended to be called in stateful validation
-  """
-  def can_spend?(output_spent, witness, raw_tx)
+  def from_db_value(%{owner: owner, currency: currency, amount: amount, output_type: output_type})
+      when is_binary(owner) and is_binary(currency) and is_integer(amount) and is_integer(output_type) do
```
Comment (InoMurko) [inline apps/omg/lib/omg/output.ex]: I don't understand these guards? who are they for? what are they protecting us from?

File: apps/omg/lib/omg/output.ex [hunk 7]
```diff
@@ -19,56 +19,75 @@ defmodule OMG.Output do
...
-  def to_db_value(output)
+  def input_pointer(%__MODULE__{}, blknum, tx_index, oindex, _, _),
+    do: Utxo.position(blknum, tx_index, oindex)
 
-  @doc """
-  Transforms into a RLP-ready structure
-  """
-  def get_data_for_rlp(output)
+  def to_db_value(%__MODULE__{owner: owner, currency: currency, amount: amount, output_type: output_type})
+      when is_binary(owner) and is_binary(currency) and is_integer(amount) and is_integer(output_type) do
```
Comment (InoMurko) [inline apps/omg/lib/omg/output.ex]: I don't understand these guards? who are they for? what are they protecting us from?

File: apps/omg/lib/omg/state/transaction.ex [hunk 8]
```diff
@@ -143,13 +143,13 @@ defprotocol OMG.State.Transaction.Protocol do
   @doc """
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/transaction.ex]: is there a more simple approach to implement what this protocol is trying to achieve?

**Pull request merged**